Doc/c-api/init.rst

```diff
@@ -1000,6 +1004,78 @@ thread, where the CPython global runtime was originally initialized.
 The only exception is if :c:func:`exec` will be called immediately
 after.
 
+.. _cautions-regarding-runtime-finalization:
+
+Cautions regarding runtime finalization
+---------------------------------------
+
+In the late stage of :term:`interpreter shutdown`, after attempting to wait for
+non-daemon threads to exit (though this can be interrupted by
+:class:`KeyboardInterrupt`) and running the :mod:`atexit` functions, the runtime
+is marked as *finalizing*: :c:func:`_Py_IsFinalizing` and
+:func:`sys.is_finalizing` return true.  At this point, only the *finalization
+thread* that initiated finalization (typically the main thread) is allowed to
+acquire the :term:`GIL`.
+
+If any thread, other than the finalization thread, attempts to acquire the GIL
+during finalization, either explicitly via :c:func:`PyGILState_Ensure`,
+:c:macro:`Py_END_ALLOW_THREADS`, :c:func:`PyEval_AcquireThread`, or
+:c:func:`PyEval_AcquireLock`, or implicitly when the interpreter attempts to
+reacquire it after having yielded it, the thread enters a permanently blocked
+state where it remains until the program exits.  In most cases this is harmless,
+but this can result in deadlock if a later stage of finalization attempts to
+acquire a lock owned by the blocked thread, or otherwise waits on the blocked
+thread.
+
+To avoid non-Python threads becoming blocked, or Python-created threads becoming
+blocked while executing C extension code, you can use
+:c:func:`PyThread_TryAcquireFinalizeBlock` and
+:c:func:`PyThread_ReleaseFinalizeBlock`.
```

Comment:
I think the wording on this is understating things.

In order to avoid threads hanging, doesn't this mean that all Python C API code and C extension code in the entire world should be using these APIs?

"you can use" implies that the user is in control of the code now found to be hanging. I don't think this will be the common case.

Reply:
(left a broader top level comment)